### PR TITLE
Fixes #1536 - Foreman should check DHCP entries on build.

### PR DESCRIPTION
**Pull request by bagasse**
http://projects.theforeman.org/issues/1536

Force foreman to rebuild dhcp entry when the host build status change to true.

### CODE REVIEW

**Comment by stbenjam** [comment]
[test]

File: app/models/concerns/orchestration/dhcp.rb [hunk 1]
```diff
@@ -93,8 +93,8 @@ def queue_dhcp_update
 
   # do we need to update our dhcp reservations
   def dhcp_update_required?
-    # IP Address / name changed
-    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet))
+    # IP Address / name changed, or 'rebuild' action is triggered.
+    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet) or (old.build? != build? and build?))
```
Comment (ohadlevy) [inline app/models/concerns/orchestration/dhcp.rb]: would a better option to query if its valid? AFAIR the record has a valid? and conflicting? methods.
this would blindly delete and recreate the reservation every time?

**Comment by bagasse** [comment]
As i seen, the code in the smart proxy, test if the record exist before adding it (https://github.com/theforeman/smart-proxy/blob/25691fff37c8ce8eaa7c62ac47a60befd6bfa329/modules/dhcp/server.rb#L119). So yes foreman will always ask the DHCP smart proxy to add the DHCP record, and the smart proxy will check the record before adding it. Because the aim of the small patch is to resynchronize the DHCP proxy with the foreman view before the build, i think it's the good way to do the thing, but i'm not familiar with the foreman logic and it's my first contribution to the project so any pointer is welcome :)

**Comment by ohadlevy** [comment]
the main difference that this will always schedule of deletion, I would add an explicit validation check and add to the queue accordingly instead?

**Comment by dLobatog** [comment]
[test] because of the latest enabled cops

**Comment by ares** [comment]
[test]

File: app/models/concerns/orchestration/dhcp.rb [hunk 2]
```diff
@@ -92,8 +92,9 @@ def queue_dhcp_update
 
   # do we need to update our dhcp reservations
   def dhcp_update_required?
-    # IP Address / name changed
-    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet))
+    # IP Address / name changed, or 'rebuild' action is triggered and DHCP record on the smart proxy is not present/identical.
+    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet) or
+                    (old.build? != build? and build? and !dhcp_record.valid?))
```
Comment (ares) [inline app/models/concerns/orchestration/dhcp.rb]: `old.build? != build? and build?`
I think is the same as
`!old.build? and build?`
which seems more readable to me

File: test/unit/orchestration/dhcp_test.rb [hunk 3]
```diff
@@ -130,6 +130,40 @@ def teardown
     assert_equal 1, h.primary_interface.queue.items.select {|x| x.action.last == :del_dhcp }.size
   end
 
+  test "when an existing host trigger a 'rebuild', its dhcp record should be updated if no dhcp record is found" do
+    Net::DHCP::Record.any_instance.stubs(:valid?).returns(false)
+    h = FactoryGirl.create(:host, :with_dhcp_orchestration)
+
+    # We need that the medium used belongs to the operating system used
```
Comment (ares) [inline test/unit/orchestration/dhcp_test.rb]: these 5 lines seems unrelated, the test passes even without them

File: test/unit/orchestration/dhcp_test.rb [hunk 4]
```diff
@@ -130,6 +130,40 @@ def teardown
...
+    assert h.valid?, h.errors.messages.to_s
+    assert_equal 2, h.queue.items.select {|x| x.action == [ h.primary_interface, :set_dhcp ] }.size
+    assert_equal 1, h.primary_interface.queue.items.select {|x| x.action.last == :del_dhcp }.size
+  end
+
+  test "when an existing host trigger a 'rebuild', its dhcp record should not be updated if valid dhcp record is found" do
+    Net::DHCP::Record.any_instance.stubs(:valid?).returns(true)
+    h = FactoryGirl.create(:host, :with_dhcp_orchestration)
+
+    # We need that the medium used belongs to the operating system used
```
Comment (ares) [inline test/unit/orchestration/dhcp_test.rb]: these 5 lines seems unrelated, the test passes even without them

**Comment by ares** [comment]
Could you please squash all commits into one? And please change the commit message from `Refs` to `Fixes`, IIUC it really fixes issue 1536

**Comment by ares** [comment]
Tested, works good, please fix minors issues from above and it's good to merge.

File: test/unit/orchestration/dhcp_test.rb (hunk 3 above)
Comment (bagasse) [inline test/unit/orchestration/dhcp_test.rb]: Could you tell me lines numbers, it's not easy to see on the interface: it show lines 133 to 137, but the first is a start of test block, one is blank, another is a comment comment.

File: test/unit/orchestration/dhcp_test.rb (hunk 4 above)
Comment (bagasse) [inline test/unit/orchestration/dhcp_test.rb]: Idem, it show me from 133 to 154.

**Comment by bagasse** [comment]
- Condition rewrited
- Tests cleaned up
- Commits squashed
- Rebased to develop

**Comment by ares** [comment]
[test]

**Comment by bagasse** [comment]
Someone can re open this issue?  It was closed by a commit message that refer a redmine ticket with the same number of this pull request.

Thank you.

File: test/factories/host_related.rb [hunk 5]
```diff
@@ -89,7 +89,7 @@ def set_nic_attributes(host, attributes, evaluator)
     end
 
     trait :with_medium do
-      medium
+      medium { FactoryGirl.create(:medium, :os_family => Operatingsystem.families[0]) }
```
Comment (ares) [inline test/factories/host_related.rb]: seems unrelated, tests are running fine for me without this
EDIT: sorry, I should have spotted sooner...

**Comment by bagasse** [comment]
Fixed.

**Comment by ares** [comment]
[test]

File: app/models/concerns/orchestration/dhcp.rb [hunk 6]
```diff
@@ -94,8 +94,9 @@ def queue_dhcp_update
 
   # do we need to update our dhcp reservations
   def dhcp_update_required?
-    # IP Address / name changed
-    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet))
+    # IP Address / name changed, or 'rebuild' action is triggered and DHCP record on the smart proxy is not present/identical.
+    return true if ((old.ip != ip) or (old.hostname != hostname) or (old.mac != mac) or (old.subnet != subnet) or
```
Comment (shlomizadok) [inline app/models/concerns/orchestration/dhcp.rb:98]: code style: `or` should be `||`, `and` should be `&&`

Comment (ares) [inline app/models/concerns/orchestration/dhcp.rb:98]: that's true, but that would have to be changed in code that is not introduced by this PR, I'd rather keep changes in this area minimal until we extract this to new objects...

**Comment by ares** [comment]
[test], seems unrelated

**Comment by ares** [comment]
Merged as https://github.com/theforeman/foreman/commit/87c6feb2227813507ae8a2911ef7f51475fa081a, thank you @bagasse!
